### ISSUE TITLE
Installation of Python3-salt break Ansible filter modules

### ISSUE THREAD

**Issue by phibid** [uyuni-project/uyuni#2258]
Uyuni release 2020.05

We are using Ansible as automation code, so we need to have Ansible and Salt to coexist on the Uyuni servers and proxies. Unfortunately there is a conflict with the Python libraries which somehow breaks the use of some YAML parsing filters like ```from_yaml```.

As example, here is a simple use of ```from_yaml``` filter, to get a dictionary from a YAML formatted string:

```yaml
---
- hosts: localhost
  gather_facts: false
  pre_tasks:
  - set_fact:
      ldap_result: "gidNumber: 602\ncn: Bruce Wayne\ngivenName: Bruce\nmail: bruce.wayne@mymail.com\nsn: Wayne\nuidNumber: 696"
  
  - name: New fact with LDAP user details
    set_fact:
      ldap_user: "{{ ldap_result|from_yaml }}"
  
  - name: CN of LDAP user
    debug:
        msg: "{{ ldap_user.cn }}"
```

The working result:

```
# cat /etc/os-release
NAME="openSUSE Leap"
VERSION="15.1"
ID="opensuse-leap"
ID_LIKE="suse opensuse"
VERSION_ID="15.1"
PRETTY_NAME="openSUSE Leap 15.1"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:leap:15.1"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
# ansible --version
ansible 2.9.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.10 (default, Jan 16 2020, 09:12:04) [GCC]
# ansible-playbook test.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **************************************************************************************************************************************************************************

TASK [set_fact] ***************************************************************************************************************************************************************************
ok: [localhost]

TASK [New fact with LDAP user details] ****************************************************************************************************************************************************
ok: [localhost]

TASK [CN of LDAP user] ********************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "Bruce Wayne"
}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
In debug mode, I can see the ```core``` filter module being correctly loaded:

```
  2884 1590532591.64539: Loading TestModule 'files' from /usr/lib/python3.6/site-packages/ansible/plugins/test/files.py
  2884 1590532591.64598: Loading TestModule 'mathstuff' from /usr/lib/python3.6/site-packages/ansible/plugins/test/mathstuff.py
  2884 1590532591.65167: Loading FilterModule 'core' from /usr/lib/python3.6/site-packages/ansible/plugins/filter/core.py (found_in_cache=True, class_only=False)
```
However after the installation of ```python3-salt``` (being a direct dependency of ```salt```), the execution of this same playbook fail:

```
# salt-call --version
salt-call 2019.2.0 (Fluorine)
# ansible-playbook test.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **************************************************************************************************************************************************************************

TASK [set_fact] ***************************************************************************************************************************************************************************
ok: [localhost]

TASK [New fact with LDAP user details] ****************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "Unexpected templating type error occurred on ({{ ldap_result|from_yaml }}): a string or stream input is required"}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

In debug mode, I can see that Ansible is failing to load the ```core``` filter module:

```
  2987 1590532711.65247: Loading TestModule 'files' from /usr/lib/python3.6/site-packages/ansible/plugins/test/files.py
  2987 1590532711.65306: Loading TestModule 'mathstuff' from /usr/lib/python3.6/site-packages/ansible/plugins/test/mathstuff.py
  2987 1590532711.65634: failing because of a type error, template data is: {{ ldap_result|from_yaml }}
```

Not sure how to go deeper on this, but if you have an idea I can perhaps help in checking more things.

**Comment by meaksh** [comment]
@phibid thanks for reporting this - could it be that when you installed `python3-salt` you also get `python3-PyYAML` or `python3-Jinja2` packages installed and that is breaking the playbook?

Could you please double check which are the versions for those packages before/after you experience the issue?

Thanks in advance!

**Comment by phibid** [comment]
I was thinking about those 2 python libraries too, but seems to me that they are using the same one:

On a fresh install of Opensuse 15.1 , all packages updated:

After the installation of Ansible:

```
# rpm -qa | egrep '(python3-PyYAML|python3-Jinja2)'
python3-PyYAML-5.1.2-lp151.2.13.1.x86_64
python3-Jinja2-2.10.1-lp151.1.1.noarch
```

After the installation of Salt (zypper in salt):

```
# rpm -qa | egrep '(python3-PyYAML|python3-Jinja2)'
python3-PyYAML-5.1.2-lp151.2.13.1.x86_64
python3-Jinja2-2.10.1-lp151.1.1.noarch
```

So not sure exactly at which level the installation of Salt broke the Ansible python libraries. Knowing that the result of the command ```rpm -ql python3-salt``` show me that all the deployed files are located in ```/usr/lib/python3.6/site-packages/salt```

The fact is that after the removal of python3-salt and salt, the Ansible playbook works again:

```
# zypper rm python3-salt
Reading installed packages...
Resolving package dependencies...

The following 2 packages are going to be REMOVED:
  python3-salt salt

2 packages to remove.
After the operation, 41.0 MiB will be freed.
Continue? [y/n/v/...? shows all options] (y): y
(1/2) Removing python3-salt-2019.2.0-lp151.5.18.1.x86_64 ............................................................................................................................[done]
(2/2) Removing salt-2019.2.0-lp151.5.18.1.x86_64 ....................................................................................................................................[done]
 
# ansible-playbook test.yml 
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **************************************************************************************************************************************************************************

TASK [set_fact] ***************************************************************************************************************************************************************************
ok: [localhost]

TASK [New fact with LDAP user details] ****************************************************************************************************************************************************
ok: [localhost]

TASK [CN of LDAP user] ********************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "Bruce Wayne"
}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

**Comment by phibid** [comment]
Does somebody had a chance to check this ? I was wondering if perhaps an issue should be created on Github Salt

**Comment by paususe** [comment]
@phibid 
Does this still happen with Uyuni 2020.09 on openSUSE Leap 15.2?

**Comment by phibid** [comment]
@paususe Unfortunately yes

```

**Comment by paususe** [comment]
@brejoc could you please take a look? It seems some packaging conflict.

@phibid are you using EPEL or some other third-party repository?

**Comment by phibid** [comment]
No, not using EPEL. Here are the repos on a freshly installed OpenSuse Leap 15.2:

```
# zypper repos
Repository priorities are without effect. All enabled repositories share the same priority.

#  | Alias                                                | Name                                        | Enabled | GPG Check | Refresh
---+------------------------------------------------------+---------------------------------------------+---------+-----------+--------
 1 | openSUSE-Leap-15.2-1                                 | openSUSE-Leap-15.2-1                        | Yes     | (r ) Yes  | Yes
 2 | repo-debug                                           | Debug Repository                            | No      | ----      | ----
 3 | repo-debug-non-oss                                   | Debug Repository (Non-OSS)                  | No      | ----      | ----
 4 | repo-debug-update                                    | Update Repository (Debug)                   | No      | ----      | ----
 5 | repo-debug-update-non-oss                            | Update Repository (Debug, Non-OSS)          | No      | ----      | ----
 6 | repo-non-oss                                         | Non-OSS Repository                          | No      | ----      | ----
 7 | repo-oss                                             | Main Repository                             | Yes     | (r ) Yes  | Yes
 8 | repo-source                                          | Source Repository                           | No      | ----      | ----
 9 | repo-source-non-oss                                  | Source Repository (Non-OSS)                 | No      | ----      | ----
10 | repo-update                                          | Main Update Repository                      | Yes     | (r ) Yes  | Yes
11 | repo-update-non-oss                                  | Update Repository (Non-Oss)                 | Yes     | (r ) Yes  | Yes
12 | susemanager:opensuse_leap15_2-x86_64                 | openSUSE Leap 15.2 (x86_64)                 | Yes     | ( p) Yes  | Yes
13 | susemanager:opensuse_leap15_2-x86_64-non-oss         | openSUSE 15.2 non oss (x86_64)              | Yes     | ( p) Yes  | Yes
14 | susemanager:opensuse_leap15_2-x86_64-non-oss-updates | openSUSE Leap 15.2 non oss Updates (x86_64) | Yes     | ( p) Yes  | Yes
15 | susemanager:opensuse_leap15_2-x86_64-updates         | openSUSE Leap 15.2 Updates (x86_64)         | Yes     | ( p) Yes  | Yes
```

The presence or not of the `python3-salt` package will condition the proper running of the Ansible playbook I mention on the description of this issue.

**Comment by brejoc** [comment]
@phibid You might have run into this issue: https://github.com/ansible/ansible/issues/70357

Ansible 2.10 should fix that. I'm not sure if you are using the saltstack connection plugin. Could this maybe deactivated? I guess this should also solve the problem for Ansible 2.9.x.

**Comment by phibid** [comment]
Good catch, this is exactly the same issue !!

To be honest, no idea if I am using the saltstack connection plugin, as all the Salt env configuration is managed by Uyuni. I think that I will wait for Ansible 2.10 to be available on OpenSuse Leap 15. Not sure how long I will have to wait.

**Comment by brejoc** [comment]
@phibid I think it should be around the corner. But I'm not very familiar with Ansible. Is it really not possible to deactivate this connection plugin somehow?

**Comment by meaksh** [comment]
FYI this issue should be fixed in Salt by https://github.com/saltstack/salt/pull/59746

With this fix, Ansible 2.9 should work together with Salt.

I'm backporting this to our Salt packages

**Comment by meaksh** [comment]
This issue should be now fixed in our Salt `products:testing` package by https://build.opensuse.org/request/show/877905

It should land in Uyuni Salt packages in the next releases, so I'm closing this issue now.

Thanks a lot @phibid for reporting this and for all the support here :+1:
